py/modbuiltins.c

```diff
@@ -225,20 +225,16 @@ STATIC mp_obj_t mp_builtin_dir(size_t n_args, const mp_obj_t *args) {
     }
 
     mp_obj_t dir = mp_obj_new_list(0, NULL);
+    mp_map_elem_t *elem;
     if (dict != NULL) {
-        for (mp_uint_t i = 0; i < dict->map.alloc; i++) {
-            if (MP_MAP_SLOT_IS_FILLED(&dict->map, i)) {
-                mp_obj_list_append(dir, dict->map.table[i].key);
-            }
+        for (size_t i = 0; (elem = mp_map_iter_next(&dict->map, &i)); ) {
+            mp_obj_list_append(dir, elem->key);
         }
     }
-    if (members != NULL) {
```

Comment:
I moved that check down into `mp_map_iter_next()` in case there were other instances that could save a `NULL` check.  It must be a compiler optimization issue, because the test passed on my build.  I'll pull it out of the inner function and put it back into `py/modbuiltins.c`.